### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by ps789**

### CODE REVIEW

File: a2c_model.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    def forward(self, x):
+        x = self.fc(x)
+        x, val = self.value(x, return_features=True)
+        logp = torch.log_softmax(self.policy(x), -1)
+        return logp, val
+
+
+class ResidualBlock(nn.Module):
+    """Following the structure of the one implemented in
```
Comment (bcakhavan) [inline a2c_model.py:34]: Helpful comment

File: envs.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        if self.prev_obs is not None:
+            # This is necesarry because the character might be dying during multiple frames
+            if not isDying(self.prev_obs.players[self.pid]) and \
+               isDying(self.obs.players[self.pid]):
+                r -= 1.0
+            
+        #     # We give a reward of -0.01 for every percent taken. The max() ensures that not reward is given when a character dies
+        #     r -= 0.01 * max(0, self.obs.players[self.pid].percent - self.prev_obs.players[self.pid].percent)
+
+        r += self.obs.players[0].y / 50 / 60
```
Comment (bcakhavan) [inline envs.py]: What does this new reward mean? Maybe comment could help?

File: train.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    optimizer.step()
+    optimizer.zero_grad()
+
+
+
+def save_model(net, optimizer, PATH):
+    torch.save({
+            'model': net,
+            'optimizer': optimizer,
+            }, PATH)
```
Comment (bcakhavan) [inline train.py]: tbh looks good to me

File: Readme.md [hunk 4]
```diff
@@ -80,11 +80,9 @@ This compiles and links the headless version of Dolphin use for training an agen
 
 ## TODO
 
-- [x] Credit all the people that made this possible
-- [x] Document all the options
-- [x] Make and test build instruction for other OS
 - [ ] Clean most of the code to make it as simple as possible (since the basis of this work is vladfi1/phillip, there is still a lot of stuff that is not used and can be removed)
 - [x] Make an example for self-play environment
+- [ ] Since issues are 1. Trajectories widely different in large state space making sample the inefficient on-policy alg dog water 2. noisy samples (improved w/ advantage to normalize) 3. sparse rewards. We want to try batch norm or dropout to regularize the network so easier to train
```
Comment (bcakhavan) [inline Readme.md:86]: Very cool todo doc

File: envs.py [hunk 5]
```diff
@@ -40,15 +42,16 @@ def embed_obs(self, obs):
     def compute_reward(self):
         r = 0.0
         if self.prev_obs is not None:
-            # This is necesarry because the character might be dying during multiple frames
+            # This is necessary because the character might be dying during multiple frames
```
Comment (bcakhavan) [inline envs.py]: spelling is important

File: envs.py [hunk 6]
```diff
@@ -40,15 +42,16 @@ def embed_obs(self, obs):
...
             if not isDying(self.prev_obs.players[self.pid]) and \
                isDying(self.obs.players[self.pid]):
                 r -= 1.0
             
-        #     # We give a reward of -0.01 for every percent taken. The max() ensures that not reward is given when a character dies
-        #     r -= 0.01 * max(0, self.obs.players[self.pid].percent - self.prev_obs.players[self.pid].percent)
+            # We give a reward of -0.01 for every percent taken and +0.01 for damage dealt. 
+            # The max() ensures that not reward is given when a character dies
+            r += -0.01 * max(0, self.obs.players[self.pid].percent - self.prev_obs.players[self.pid].percent) + 0.01 * self.obs.players[opponent_pid].percent
```
Comment (bcakhavan) [inline envs.py]: qq wouldn't you need a difference here to make it be incentivized by opponent losing percentages? np.abs(self.prev_obs.players[opponent_pid].percent - self.obs.players[opponent_pid].percent)?

File: train.py [hunk 7]
```diff
@@ -39,6 +39,7 @@ def train(params, net, optimizer, env):
 
 
 def gather_rollout(params, net, env, obs, prnt = False):
+    """ Obs |> net -> action, values. Action |> env.step -> (reward, taken action (sampled from action_probs), action_probs, values ), obs"""
```
Comment (bcakhavan) [inline train.py:93]: love the comments ty ty

File: train.py [hunk 8]
```diff
@@ -13,7 +13,8 @@ def train(params, net, optimizer, env):
     n_save = 50000
     start_time = time.time()
 
-    while total_steps < params.total_steps:
+    # while total_steps < params.total_steps:
+    while True:
```
Comment (bcakhavan) [inline train.py:62]: qq how do we break out of the loop

File: envs.py [hunk 9]
```diff
@@ -52,8 +52,12 @@ def compute_reward(self):
                isDying(self.obs.players[self.pid]):
                 r -= 1.0
             
+            if not isDying(self.prev_obs.players[opponent_pid]) and \
+               isDying(self.obs.players[opponent_pid]):
+                r += 1.0
+
         #     # We give a reward of -0.01 for every percent taken. The max() ensures that not reward is given when a character dies
-            r += -0.01 * max(0, self.obs.players[self.pid].percent - self.prev_obs.players[self.pid].percent) + 0.01 * (self.obs.players[opponent_pid].percent)
+            r += -0.01 * max(0, self.obs.players[self.pid].percent - self.prev_obs.players[self.pid].percent) + 0.01 * (self.obs.players[opponent_pid].percent - self.prev_obs.players[opponent_pid].percent)
```
Comment (bcakhavan) [inline envs.py]: Oh u fixed it yay. 

qq is it possible that the terms in the second element of the max() are reversed? Shouldn't it be prev percent - current percent + if that's positive then *-0.01? I could be reading it wrong

File: train.py (hunk 8 above)
Comment (bayoumi17m) [reply train.py:62]: This^

File: train.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    obs = env.reset()
+    total_steps = 0
+    n_save = 50000
+    start_time = time.time()
+
+    # while total_steps < params.total_steps:
+    while True:
+        print("Total steps:", total_steps)
+
+        # Gathering rollouts: for 600 steps, run the network in the environment without updating network
```
Comment (bayoumi17m) [inline train.py:65]: Why 600 steps? Should this be a hyperparameter that is controlled?

File: train.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
+import torch
+import torch.nn as nn
+from torch.distributions import Categorical
+from statistics import mean, stdev
+import time
+import pandas as pd
+import numpy as np
+
+def train(params, net, optimizer, env):
```
Comment (bayoumi17m) [inline train.py]: Doc strings?

File: play.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import os
+from envs import GoHighEnv
+import atexit
+import platform
+
+checkpoint = torch.load(os.path.join("checkpoints", "agent.ckpt"))
+
+net = checkpoint["model"]
+
+options = dict(
```
Comment (bayoumi17m) [inline play.py:12]: Minor thing: It would be better to have this as a configuration file that is read in so we can specify different configurations easily.

File: envs.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+        env = GoHighEnv(frame_limit=frame_limit, options=options)
+        return env
+    return _init
+
+
+def GoHighEnvVec(num_envs, frame_limit=1e9, options={}):
+    return SubprocVecEnv([make_env(frame_limit=frame_limit, options=options) for _ in range(num_envs)])
+
+
+class GoHighEnv(BaseEnv):
```
Comment (bayoumi17m) [inline envs.py:18]: What is GoHigh?

File: a2c_model.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
+import torch
+import torch.nn as nn
+
+
+class Actor(nn.Module):
```
Comment (bayoumi17m) [inline a2c_model.py:5]: Why not agent lol. This seems like an Actor Critic.

File: a2c_model.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, obs_dim, action_dim, hidden_dim = 256):
+        """initializes embedding, value network (ResNet, state -> value), and policy network (state -> action) """
+        super(Actor, self).__init__()
+        # self.fc = nn.Linear(obs_dim, hidden_dim)
+        self.fc = nn.Linear(obs_dim * 2, hidden_dim)
+        self.value = ResNet(hidden_dim, 1, 2, output_dim=1)
+        self.policy = nn.Linear(hidden_dim, action_dim)
```
Comment (bayoumi17m) [inline a2c_model.py]: Why is the value network so different than the policy network?

**Review by bayoumi17m — COMMENTED** [review]
Overall, things are pretty clear since they are simple but  everything needs to be much better documented. Additionally you should purge your checkpoints from the REPO.

**Review by MelindaFang-code — COMMENTED** [review]
What model are you actually using though? Are you using actor-critic? Cause I thought A2C (advantage actor-critic) is a type of actor-critic algorithm

File: train.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    ep_rewards = [0.] * params.num_workers
+    t = time.time()
+    for _ in range(params.rollout_steps):
+        obs = torch.tensor(obs)
+        logps, values = net(obs)
+        actions = Categorical(logits=logps).sample()
+
+        obs, rewards, dones, _ = env.step(actions.numpy())
+
+        for i, done in enumerate(dones):
```
Comment (MelindaFang-code) [inline train.py:104]: what do you do when done = 1 in this case?

File: train.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+    if prnt:
+        to_print = {"time": round(time.time() - t, 3), "reward_mean": round(mean(ep_rewards), 3), "reward_std":round(stdev(ep_rewards), 3)}
+        return steps, obs, to_print
+        print(to_print)
+    return steps, obs, None
+
+
+def process_rollout(params, steps):
+    # bootstrap discounted returns with final value estimates
```
Comment (MelindaFang-code) [inline train.py:117]: clear documentation of each function 👍
